src/process/unix/SDL_process.c

```diff
@@ -0,0 +1,314 @@
+        char **mutable_env = env ? dupstrlist(env) : NULL;
+
+        // We are in a new process; don't bother freeing because either exec*() or exit() will succeed
+
+        if (!mutable_args || (env && !mutable_env)) {
+            fprintf(stderr, "Could not clone args/env: %s\n", SDL_GetError());
+            exit(1);
+        }
+
+        if (env) {
+            execve(args[0], mutable_args, mutable_env);
+        } else {
+            execv(args[0], mutable_args);
+        }
+
+        // If this is reached, execv() failed
+
+        if (flags & SDL_PROCESS_ERRORS_TO_STDERR) {
+            fprintf(stderr, "Could not execv/execve(): %s\n", strerror(errno));
+        }
+
+        exit(1);
+    } else {
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[READ_END]);
+        }
+
+        if (flags & SDL_PROCESS_STDOUT) {
+            close(process->stdout_pipe[WRITE_END]);
+        }
+
+        if (flags & SDL_PROCESS_STDERR) {
+            close(process->stderr_pipe[WRITE_END]);
+        }
+
+        return process;
+    }
+}
+
+int SDL_WriteProcess(SDL_Process *process, const void *buffer, int size)
+{
+    if (!process) {
+        SDL_SetError("Attempt to call SDL_WriteProcess() with NULL process");
+        return -1;
+    }
+
+    if (!(process->flags & SDL_PROCESS_STDIN)) {
+        SDL_SetError("Cannot read from process created without SDL_PROCESS_STDIN");
+        return -1;
+    }
+
+    int ret = write(process->stdin_pipe[WRITE_END], buffer, size);
+
+    if (ret < 0) {
+        SDL_SetError("Could not write(): %s", strerror(errno));
+    }
+
+    return ret;
+}
+
+int SDL_ReadProcess(SDL_Process *process, void *buffer, int size)
+{
+    if (!process) {
+        SDL_SetError("Attempt to call SDL_ReadProcess() with NULL process");
+        return -1;
+    }
+
+    if (!(process->flags & SDL_PROCESS_STDOUT)) {
+        SDL_SetError("Cannot read from process created without SDL_PROCESS_STDOUT");
+        return -1;
+    }
+
+    int ret = read(process->stdout_pipe[READ_END], buffer, size);
+
+    if (ret < 0) {
+        SDL_SetError("Could not read(): %s", strerror(errno));
+    }
+
+    return ret;
+}
+
+int SDL_ReadErrProcess(SDL_Process *process, void *buffer, int size)
+{
+    if (!process) {
+        SDL_SetError("Attempt to call SDL_ReadErrProcess() with NULL process");
+        return -1;
+    }
+
+    if (!(process->flags & SDL_PROCESS_STDERR)) {
+        SDL_SetError("Cannot read from process created without SDL_PROCESS_STDERR");
+        return -1;
+    }
+
+    int ret = read(process->stderr_pipe[READ_END], buffer, size);
+
+    if (ret < 0) {
+        SDL_SetError("Could not read(): %s", strerror(errno));
+    }
+
+    return ret;
+}
+
+SDL_bool SDL_KillProcess(SDL_Process *process, SDL_bool force)
+{
+    if (!process) {
+        SDL_SetError("Attempt to call SDL_KillProcess() with NULL process");
+        return -1;
+    }
+
+    int ret = kill(process->pid, force ? SIGKILL : SIGTERM);
```

Comment:
If you have already called `waitpid()` for the process's PID, this could kill an unrelated process if the PID has been reused. Ideally the call to `waitpid()` should remember that it has done so, and make `SDL_KillProcess()` not do anything.

Reply:
Related to that (because it updates mutable state in the struct), what's the intended thread-safety of a `SDL_Process`? As implemented here, I think it's something like: each `SDL_Process` can only be used by one thread at a time, and concurrent use from more than one thread is undefined behaviour.

Reply:
I haven't yet planned the thread safety of SDL_process, because I wanted to open the PR early with the simplest implementation. I can see after the other PR is merged about what can be made thread-safe, if it can be useful.

Reply:
Needing thread-safety often has a significant effect on the design that's needed, so I find that it's often best to have an idea of where you're going from the beginning - even if that idea is a simplifying assumption like "not thread-safe, use each object from one thread only" which doesn't put many constraints on the design/implementation (you can't use global/static variables but otherwise everything can be simple).